File: doc/examples/DocExamplesVB/ProductGuide/Authentication.vb
Imports Stormpath.SDK
Imports Stormpath.SDK.Account
Imports Stormpath.SDK.Api
Imports Stormpath.SDK.Application
Imports Stormpath.SDK.Auth
Imports Stormpath.SDK.Client
Imports Stormpath.SDK.Directory
Imports Stormpath.SDK.Error
Imports Stormpath.SDK.Oauth
Imports Stormpath.SDK.Provider

Public Class Authentication

    Private client As IClient = Nothing

    Public Async Function LoginAttempt() As Task
#Region "code/vbnet/authentication/login_attempt_req.vb"
        Try
            Dim app = Await client.GetApplicationAsync(
                "https://api.stormpath.com/v1/applications/$YOUR_APPLICATION_ID")

            Dim loginResult = Await app.AuthenticateAccountAsync(
                "han@newrepublic.gov", "first2shoot!")

            Dim accountDetails = Await loginResult.GetAccountAsync()

        Catch rex As ResourceException
            ' Bad login credentials!
            Console.WriteLine("Error logging in. " + rex.Message)
        End Try
#End Region
    End Function

    Public Async Function LoginAttemptExpanded() As Task
        Dim app As IApplication = Nothing

#Region "code/vbnet/authentication/login_attempt_req_expand_account.vb"
        Dim loginRequest = New UsernamePasswordRequestBuilder() _
            .SetUsernameOrEmail("han@newrepublic.gov") _
            .SetPassword("first2shoot!") _
            .Build()

        Dim loginResult = Await app.AuthenticateAccountAsync(
            loginRequest,
            Function(opt) opt.Expand(Function(result) result.GetAccount()))
#End Region
    End Function


    Public Async Function CustomDataAccountLink() As Task
        Dim account As IAccount = Nothing

#Region "code/vbnet/authentication/customdata_accountlink.vb"
        Dim customData = Await account.GetCustomDataAsync()

        customData.Put("https://api.stormpath.com/v1/accounts/3fLduLKlEx")
        Await customData.SaveAsync()
#End Region
    End Function

    Public Async Function GetAccountStoreMappings() As Task
#Region "code/vbnet/authentication/get_asm_req.vb"
        Dim app = Await client.GetApplicationAsync(
            "https://api.stormpath.com/v1/applications/5nan67mWrYrBmLGu7nGurh")

        Dim accountStoreMappings = Await app _
            .GetAccountStoreMappings() _
            .ToListAsync()
#End Region
    End Function

    Public Async Function CreateDefaultAccountStoreMapping() As Task
#Region "code/vbnet/authentication/create_asm.vb"
        Dim application = Await client.GetApplicationAsync(
            "https://api.stormpath.com/v1/applications/1gk4Dxzi6o4PbdleXaMPL")

        Dim directory = Await client.GetDirectoryAsync(
            "https://api.stormpath.com/v1/directories/2SKhstu8PlaekcaEXampLE")

        Dim newMapping = client.Instantiate(Of IApplicationAccountStoreMapping)() _
            .SetApplication(application) _
            .SetAccountStore(directory) _
            .SetListIndex(0) _
            .SetDefaultAccountStore(True) _
            .SetDefaultGroupStore(True)

        Await application.CreateAccountStoreMappingAsync(newMapping)
#End Region
    End Function

    Public Async Function UpdateExistingMappingListIndex() As Task
        Dim existingMapping As IApplicationAccountStoreMapping = Nothing

#Region "code/vbnet/authentication/change_login_priority.vb"
        existingMapping.SetListIndex(0)
        Await existingMapping.SaveAsync()
#End Region
    End Function

    Public Async Function UpdateExistingMappingDefaultStores() As Task
        Dim existingMapping As IApplicationAccountStoreMapping = Nothing

#Region "code/vbnet/authentication/change_default_stores.vb"
        existingMapping.SetDefaultAccountStore(True)
        existingMapping.SetDefaultGroupStore(True)
        Await existingMapping.SaveAsync()
#End Region
    End Function

    Public Async Function GetOauthPolicy() As Task
        Dim app As IApplication = Nothing

#Region "code/vbnet/authentication/oauth_policy.vb"
        Dim oauthPolicy = Await app.GetOauthPolicyAsync()
#End Region
    End Function

    Public Async Function UpdateOauthPolicyTtls() As Task
        Dim oauthPolicy As IOauthPolicy = Nothing

#Region "code/vbnet/authentication/update_oauth_ttl_req.vb"
        oauthPolicy.SetAccessTokenTimeToLive(TimeSpan.FromMinutes(30))
        oauthPolicy.SetRefreshTokenTimeToLive(TimeSpan.FromDays(7))
        Await oauthPolicy.SaveAsync()
#End Region
    End Function

    Public Async Function OauthPasswordGrantFlow() As Task
        Dim app As IApplication = Nothing

#Region "code/vbnet/authentication/generate_oauth_token_req.vb"
        Dim passwordGrantRequest = OauthRequests.NewPasswordGrantRequest() _
            .SetLogin("han@newrepublic.gov") _
            .SetPassword("first2shoot!") _
            .Build()

        Dim grantResponse = Await app.NewPasswordGrantAuthenticator() _
            .AuthenticateAsync(passwordGrantRequest)
#End Region
    End Function

    Public Async Function OauthValidateAccessToken() As Task
        Dim app As IApplication = Nothing
        Dim access_token As String = Nothing

#Region "code/vbnet/authentication/validate_oauth_token.vb"
        Dim validationRequest = OauthRequests.NewJwtAuthenticationRequest() _
            .SetJwt(access_token) _
            .Build()

        Dim accessToken = Await app.NewJwtAuthenticator() _
            .AuthenticateAsync(validationRequest)
#End Region
    End Function

    Public Async Function OauthValidateAccessTokenLocally() As Task
        Dim app As IApplication = Nothing
        Dim access_token As String = Nothing

#Region "code/vbnet/authentication/validate_oauth_token_local.vb"
        Dim validationRequest = OauthRequests.NewJwtAuthenticationRequest() _
            .SetJwt(access_token) _
            .Build()

        Dim accessToken = Await app.NewJwtAuthenticator() _
            .WithLocalValidation() _
            .AuthenticateAsync(validationRequest)
#End Region
    End Function


    Public Async Function OauthRefreshAccessToken() As Task
        Dim app As IApplication = Nothing
        Dim refresh_token As String = Nothing

#Region "code/vbnet/authentication/refresh_access_token_req.vb"
        Dim refreshGrantRequest = OauthRequests.NewRefreshGrantRequest() _
            .SetRefreshToken(refresh_token) _
            .Build()

        Dim grantResponse = Await app.NewRefreshGrantAuthenticator() _
            .AuthenticateAsync(refreshGrantRequest)
#End Region
    End Function


    Public Async Function GetTokensForAccount() As Task
        Dim account As IAccount = Nothing
        Dim access_token As String = Nothing
        Dim refresh_token As String = Nothing

#Region "code/vbnet/authentication/get_access_tokens.vb"
        Dim allAccessTokens = Await account.GetAccessTokens().ToListAsync()
        Dim allRefreshTokens = Await account.GetRefreshTokens().ToListAsync()

        Dim accessTokenToDelete = allAccessTokens.Where(Function(at) at.Jwt = access_token)
        Dim refreshTokenToDelete = allRefreshTokens.Where(Function(rt) rt.Jwt = refresh_token)
#End Region
    End Function

    Public Async Function GetTokensForAccountForApplication() As Task
        Dim account As IAccount = Nothing

#Region "code/vbnet/authentication/get_access_tokens_for_app.vb"
        Dim allAccessTokens = Await account.GetAccessTokens() _
            .Where(Function(at) at.ApplicationHref = "https://api.stormpath.com/v1/applications/1gk4Dxzi6o4PbdlBVa6tfR") _
            .ToListAsync()
#End Region
    End Function

    Public Async Function DeleteTokens() As Task
        Dim accessTokenToDelete As IAccessToken = Nothing
        Dim refreshTokenToDelete As IRefreshToken = Nothing

#Region "code/vbnet/authentication/delete_user_access_tokens_req.vb"
        Await accessTokenToDelete.DeleteAsync()
        Await refreshTokenToDelete.DeleteAsync()
#End Region
    End Function

    Public Async Function CreateGoogleDirectory() As Task
#Region "code/vbnet/authentication/create_directory_google.vb"
        '# Imports Stormpath.SDK.Provider

        Dim googleDirectory = client.Instantiate(Of IDirectory)()
        googleDirectory.SetName("My Google Directory")

        Await client.CreateDirectoryAsync(
            googleDirectory,
            Function(options) options.ForProvider(client _
                                                  .Providers() _
                                                  .Google() _
                                                  .Builder() _
                                                  .SetClientId("YOUR_GOOGLE_CLIENT_ID") _
                                                  .SetClientSecret("YOUR_GOOGLE_CLIENT_SECRET") _
                                                  .SetRedirectUri("YOUR_GOOGLE_REDIRECT_URI") _
                                                  .Build()))
#End Region
    End Function

    Public Async Function GetGoogleAccountWithCode() As Task
        Dim app As IApplication = Nothing
        Dim code As String = Nothing

#Region "code/vbnet/authentication/create_account_google_providerdata_code.vb"
        Dim request = client.Providers() _
            .Google() _
            .Account() _
            .SetCode(code) _
            .Build()
        Dim result = Await app.GetAccountAsync(request)
#End Region
    End Function

    Public Async Function GetGoogleAccountWithAccessToken() As Task
        Dim app As IApplication = Nothing
        Dim access_token As String = Nothing

#Region "code/vbnet/authentication/create_account_google_providerdata_access_token.vb"
        Dim request = client.Providers() _
            .Google() _
            .Account() _
            .SetAccessToken(access_token) _
            .Build()
        Dim result = Await app.GetAccountAsync(request)
#End Region
    End Function

    Public Async Function CreateFacebookDirectory() As Task
#Region "code/vbnet/authentication/create_directory_fb.vb"
        '# Imports Stormpath.SDK.Provider

        Dim facebookDirectory = client.Instantiate(Of IDirectory)()
        facebookDirectory.SetName("My Facebook Directory")

        Await client.CreateDirectoryAsync(
            facebookDirectory,
            Function(options) options.ForProvider(client _
                                                  .Providers() _
                                                  .Facebook() _
                                                  .Builder() _
                                                  .SetClientId("YOUR_FACEBOOK_APP_ID") _
                                                  .SetClientSecret("YOUR_FACEBOOK_APP_SECRET") _
                                                  .Build()))
#End Region
    End Function

    Public Async Function GetFacebookAccountWithAccessToken() As Task
        Dim app As IApplication = Nothing
        Dim access_token As String = Nothing

#Region "code/vbnet/authentication/create_account_fb_providerdata_access_token.vb"
        Dim request = client.Providers() _
            .Facebook() _
            .Account() _
            .SetAccessToken(access_token) _
            .Build()
        Dim result = Await app.GetAccountAsync(request)
#End Region
    End Function


    Public Async Function CreateGithubkDirectory() As Task
#Region "code/vbnet/authentication/create_directory_github.vb"
        '# Imports Stormpath.SDK.Provider

        Dim githubDirectory = client.Instantiate(Of IDirectory)()
        githubDirectory.SetName("My Github Directory")

        Await client.CreateDirectoryAsync(
            githubDirectory,
            Function(options) options.ForProvider(client _
                                                  .Providers() _
                                                  .Github() _
                                                  .Builder() _
                                                  .SetClientId("YOUR_GITHUB_CLIENT_ID") _
                                                  .SetClientSecret("YOUR_GITHUB_CLIENT_SECRET") _
                                                  .Build()))
#End Region
    End Function

    Public Async Function GetGithubAccountWithAccessToken() As Task
        Dim app As IApplication = Nothing
        Dim access_token As String = Nothing

#Region "code/vbnet/authentication/create_account_github_providerdata_access_token.vb"
        Dim request = client.Providers() _
            .Github() _
            .Account() _
            .SetAccessToken(access_token) _
            .Build()
        Dim result = Await app.GetAccountAsync(request)
#End Region
    End Function

    Public Async Function CreateLinkedInDirectory() As Task
#Region "code/vbnet/authentication/create_directory_linkedin.vb"
        '# Imports Stormpath.SDK.Provider

        Dim linkedInDirectory = client.Instantiate(Of IDirectory)()
        linkedInDirectory.SetName("My LinkedIn Directory")

        Await client.CreateDirectoryAsync(
            linkedInDirectory,
            Function(options) options.ForProvider(client _
                                                  .Providers() _
                                                  .LinkedIn() _
                                                  .Builder() _
                                                  .SetClientId("YOUR_LINKEDIN_APP_ID") _
                                                  .SetClientSecret("YOUR_LINKEDIN_APP_SECRET") _
                                                  .Build()))
#End Region
    End Function

    Public Async Function GetLinkedInAccountWithAccessToken() As Task
        Dim app As IApplication = Nothing
        Dim access_token As String = Nothing

#Region "code/vbnet/authentication/create_account_linkedin_providerdata_access_token.vb"
        Dim request = client.Providers() _
            .LinkedIn() _
            .Account() _
            .SetAccessToken(access_token) _
            .Build()
        Dim result = Await app.GetAccountAsync(request)
#End Region
    End Function

    Public Async Function GetSamlPolicy() As Task
        Dim app As IApplication = Nothing

#Region "code/vbnet/authentication/saml_policy_example.vb"
        Dim samlPolicy = Await app.GetSamlPolicyAsync()

        Dim samlServiceProvider = Await samlPolicy.GetSamlServiceProviderAsync()
        ' samlServiceProvider.Href is the SAML Service Provider URL

        Dim ssoInitiationEndpoint = Await samlServiceProvider.GetSsoInitiationEndpointAsync()
        ' ssoInitiationEndpoint.Href is the SSO Initiation Endpoint
#End Region
    End Function

    Public Async Function CreateApiKeyForAccount() As Task
        Dim account As IAccount = Nothing

#Region "code/vbnet/authentication/create_apikey_req.vb"
        Dim newApiKey = Await account.CreateApiKeyAsync()
#End Region
    End Function

    Public Async Function DeleteApiKey() As Task
        Dim app As IApplication = Nothing
        Dim api_key_id As String = Nothing

#Region "code/vbnet/authentication/delete_apikey.vb"
        Dim apiKey = Await app.GetApiKeyAsync(api_key_id)
        Await apiKey.DeleteAsync()
#End Region
    End Function

    Public Async Function DisableApiKey() As Task
        Dim app As IApplication = Nothing
        Dim api_key_id As String = Nothing

#Region "code/vbnet/authentication/disable_apikey.vb"
        Dim apiKey = Await app.GetApiKeyAsync(api_key_id)
        apiKey.SetStatus(ApiKeyStatus.Disabled)
        Await apiKey.SaveAsync()
#End Region
    End Function

    Public Async Function AuthenticateBasicApiRequest() As Task
        Dim app As IApplication = Nothing
        Dim api_key_id As String = Nothing
        Dim api_key_secret As String = Nothing

#Region "code/vbnet/authentication/authenticate_basic_req.vb"
        Dim apiKeyAuthRequest = New ApiKeyRequestBuilder() _
            .SetId(api_key_id) _
            .SetSecret(api_key_secret) _
            .Build()

        Dim result = Await app.AuthenticateAccountAsync(apiKeyAuthRequest)
        Dim account = Await result.GetAccountAsync()
#End Region
    End Function

End Class

File: doc/examples/DocExamples/ProductGuide/Authentication.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Stormpath.SDK;
using Stormpath.SDK.Account;
using Stormpath.SDK.Api;
using Stormpath.SDK.Application;
using Stormpath.SDK.Auth;
using Stormpath.SDK.Client;
using Stormpath.SDK.Directory;
using Stormpath.SDK.Error;
using Stormpath.SDK.Oauth;
using Stormpath.SDK.Provider;

namespace examples.ProductGuide
{
    public class Authentication
    {
        IClient client = null;

        public async Task LoginAttempt()
        {
            #region code/csharp/authentication/login_attempt_req.cs
            try
            {
                var app = await client.GetApplicationAsync(
                    "https://api.stormpath.com/v1/applications/$YOUR_APPLICATION_ID");

                var loginResult = await app.AuthenticateAccountAsync(
                    "han@newrepublic.gov", "first2shoot!");

                var accountDetails = await loginResult.GetAccountAsync();
            }
            catch (ResourceException rex)
            {
                // Bad login credentials!
                Console.WriteLine("Error logging in. " + rex.Message);
            }
            #endregion
        }

        public async Task LoginAttemptExpanded()
        {
            IApplication app = null;

            #region code/csharp/authentication/login_attempt_req_expand_account.cs
            var loginRequest = new UsernamePasswordRequestBuilder()
                .SetUsernameOrEmail("han@newrepublic.gov")
                .SetPassword("first2shoot!")
                .Build();

            var loginResult = await app.AuthenticateAccountAsync(
                loginRequest,
                opt => opt.Expand(result => result.GetAccount()));
            #endregion
        }

        public async Task CustomDataAccountLink()
        {
            IAccount account = null;

            #region code/csharp/authentication/customdata_accountlink.cs
            var customData = await account.GetCustomDataAsync();

            customData.Put("https://api.stormpath.com/v1/accounts/3fLduLKlEx");
            await customData.SaveAsync();
            #endregion
        }

        public async Task GetAccountStoreMappings()
        {
            #region code/csharp/authentication/get_asm_req.cs
            var app = await client.GetApplicationAsync(
                "https://api.stormpath.com/v1/applications/5nan67mWrYrBmLGu7nGurh");

            var accountStoreMappings = await app
                .GetAccountStoreMappings()
                .ToListAsync();
            #endregion
        }

        public async Task CreateDefaultAccountStoreMapping()
        {
            #region code/csharp/authentication/create_asm.cs
            var application = await client.GetApplicationAsync(
                "https://api.stormpath.com/v1/applications/1gk4Dxzi6o4PbdleXaMPL");

            var directory = await client.GetDirectoryAsync(
                "https://api.stormpath.com/v1/directories/2SKhstu8PlaekcaEXampLE");

            var newMapping = client.Instantiate<IApplicationAccountStoreMapping>()
                .SetApplication(application)
                .SetAccountStore(directory)
                .SetListIndex(0)
                .SetDefaultAccountStore(true)
                .SetDefaultGroupStore(true);

            await application.CreateAccountStoreMappingAsync(newMapping);
            #endregion
        }

        public async Task UpdateExistingMappingListIndex()
        {
            IApplicationAccountStoreMapping existingMapping = null;

            #region code/csharp/authentication/change_login_priority.cs
            existingMapping.SetListIndex(0);
            await existingMapping.SaveAsync();
            #endregion
        }

        public async Task UpdateExistingMappingDefaultStore()
        {
            IApplicationAccountStoreMapping existingMapping = null;

            #region code/csharp/authentication/change_default_stores.cs
            existingMapping.SetDefaultAccountStore(true);
            existingMapping.SetDefaultGroupStore(true);
            await existingMapping.SaveAsync();
            #endregion
        }

        public async Task GetOauthPolicy()
        {
            IApplication app = null;

            #region code/csharp/authentication/oauth_policy.cs
            var oauthPolicy = await app.GetOauthPolicyAsync();
            #endregion
        }

        public async Task UpdateOauthPolicyTtls()
        {
            IOauthPolicy oauthPolicy = null;

            #region code/csharp/authentication/update_oauth_ttl_req.cs
            oauthPolicy.SetAccessTokenTimeToLive(TimeSpan.FromMinutes(30));
            oauthPolicy.SetRefreshTokenTimeToLive(TimeSpan.FromDays(7));
            await oauthPolicy.SaveAsync();
            #endregion
        }

        public async Task OauthPasswordGrantFlow()
        {
            IApplication app = null;

            #region code/csharp/authentication/generate_oauth_token_req.cs
            var passwordGrantRequest = OauthRequests.NewPasswordGrantRequest()
                .SetLogin("han@newrepublic.gov")
                .SetPassword("first2shoot!")
                .Build();

            var grantResponse = await app.NewPasswordGrantAuthenticator()
                .AuthenticateAsync(passwordGrantRequest);
            #endregion
        }

        public async Task OauthValidateAccessToken()
        {
            IApplication app = null;
            string access_token = null;

            #region code/csharp/authentication/validate_oauth_token_sp_req.cs
            var validationRequest = OauthRequests.NewJwtAuthenticationRequest()
                .SetJwt(access_token)
                .Build();

            var accessToken = await app.NewJwtAuthenticator()
                .AuthenticateAsync(validationRequest);
            #endregion
        }

        public async Task OauthValidateAccessTokenLocally()
        {
            IApplication app = null;
            string access_token = null;

            #region code/csharp/authentication/validate_oauth_token_local.cs
            var validationRequest = OauthRequests.NewJwtAuthenticationRequest()
                .SetJwt(access_token)
                .Build();

            var accessToken = await app.NewJwtAuthenticator()
                .WithLocalValidation()
                .AuthenticateAsync(validationRequest);
            #endregion
        }

        public async Task OauthRefreshAccessToken()
        {
            IApplication app = null;
            string refresh_token = null;

            #region code/csharp/authentication/refresh_access_token_req.cs
            var refreshGrantRequest = OauthRequests.NewRefreshGrantRequest()
                .SetRefreshToken(refresh_token)
                .Build();

            var grantResponse = await app.NewRefreshGrantAuthenticator()
                .AuthenticateAsync(refreshGrantRequest);
            #endregion
        }

        public async Task GetTokensForAccount()
        {
            IAccount account = null;
            string access_token = null;
            string refresh_token = null;

            #region code/csharp/authentication/get_access_tokens.cs
            var allAccessTokens = await account.GetAccessTokens().ToListAsync();
            var allRefreshTokens = await account.GetRefreshTokens().ToListAsync();

            var accessTokenToDelete = allAccessTokens.Where(at => at.Jwt == access_token);
            var refreshTokenToDelete = allRefreshTokens.Where(rt => rt.Jwt == refresh_token);
            #endregion
        }

        public async Task GetTokensForAccountForApplication()
        {
            IAccount account = null;

            #region code/csharp/authentication/get_access_tokens_for_app.cs
            var allAccessTokens = await account.GetAccessTokens()
                .Where(at => at.ApplicationHref == "https://api.stormpath.com/v1/applications/1gk4Dxzi6o4PbdlBVa6tfR")
                .ToListAsync();
            #endregion
        }

        public async Task DeleteTokens()
        {
            IAccessToken accessTokenToDelete = null;
            IRefreshToken refreshTokenToDelete = null;

            #region code/csharp/authentication/delete_user_access_tokens_req.cs
            await accessTokenToDelete.DeleteAsync();
            await refreshTokenToDelete.DeleteAsync();
            #endregion
        }

        public async Task CreateGoogleDirectory()
        {
            #region code/csharp/authentication/create_directory_google.cs
            //# using Stormpath.SDK.Provider;

            var googleDirectory = client.Instantiate<IDirectory>();
            googleDirectory.SetName("My Google Directory");

            await client.CreateDirectoryAsync(googleDirectory,
                options => options.ForProvider(
                    client.Providers()
                        .Google()
                        .Builder()
                        .SetClientId("YOUR_GOOGLE_CLIENT_ID")
                        .SetClientSecret("YOUR_GOOGLE_CLIENT_SECRET")
                        .SetRedirectUri("YOUR_GOOGLE_REDIRECT_URI")
                        .Build()
                    ));
            #endregion
        }

        public async Task GetGoogleAccountWithCode()
        {
            IApplication app = null;
            string code = null;

            #region code/csharp/authentication/create_account_google_providerdata_code.cs
            var request = client.Providers()
                .Google()
                .Account()
                .SetCode(code)
                .Build();
            var result = await app.GetAccountAsync(request);
            #endregion
        }

        public async Task GetGoogleAccountWithAccessToken()
        {
            IApplication app = null;
            string access_token = null;

            #region code/csharp/authentication/create_account_google_providerdata_access_token.cs
            var request = client.Providers()
                .Google()
                .Account()
                .SetAccessToken(access_token)
                .Build();
            var result = await app.GetAccountAsync(request);
            #endregion
        }

        public async Task CreateFacebookDirectory()
        {
            #region code/csharp/authentication/create_directory_fb.cs
            //# using Stormpath.SDK.Provider;

            var facebookDirectory = client.Instantiate<IDirectory>();
            facebookDirectory.SetName("My Facebook Directory");

            await client.CreateDirectoryAsync(facebookDirectory,
                options => options.ForProvider(
                    client.Providers()
                        .Facebook()
                        .Builder()
                        .SetClientId("YOUR_FACEBOOK_APP_ID")
                        .SetClientSecret("YOUR_FACEBOOK_APP_SECRET")
                        .Build()
                    ));
            #endregion
        }

        public async Task GetFacebookAccountWithAccessToken()
        {
            IApplication app = null;
            string access_token = null;
            
            #region code/csharp/authentication/create_account_fb_providerdata_access_token.cs
            var request = client.Providers()
                .Facebook()
                .Account()
                .SetAccessToken(access_token)
                .Build();
            var result = await app.GetAccountAsync(request);
            #endregion
        }

        public async Task CreateGithubkDirectory()
        {
            #region code/csharp/authentication/create_directory_github.cs
            //# using Stormpath.SDK.Provider;

            var githubDirectory = client.Instantiate<IDirectory>();
            githubDirectory.SetName("My Github Directory");

            await client.CreateDirectoryAsync(githubDirectory,
                options => options.ForProvider(
                    client.Providers()
                        .Github()
                        .Builder()
                        .SetClientId("YOUR_GITHUB_CLIENT_ID")
                        .SetClientSecret("YOUR_GITHUB_CLIENT_SECRET")
                        .Build()
                    ));
            #endregion
        }

        public async Task GetGithubAccountWithAccessToken()
        {
            IApplication app = null;
            string access_token = null;

            #region code/csharp/authentication/create_account_github_providerdata_access_token.cs
            var request = client.Providers()
                .Github()
                .Account()
                .SetAccessToken(access_token)
                .Build();
            var result = await app.GetAccountAsync(request);
            #endregion
        }

        public async Task CreateLinkedInDirectory()
        {
            #region code/csharp/authentication/create_directory_linkedin.cs
            //# using Stormpath.SDK.Provider;

            var linkedInDirectory = client.Instantiate<IDirectory>();
            linkedInDirectory.SetName("My LinkedIn Directory");

            await client.CreateDirectoryAsync(linkedInDirectory,
                options => options.ForProvider(
                    client.Providers()
                        .LinkedIn()
                        .Builder()
                        .SetClientId("YOUR_LINKEDIN_APP_ID")
                        .SetClientSecret("YOUR_LINKEDIN_APP_SECRET")
                        .Build()
            #endregion
        }

        public async Task GetLinkedInAccountWithAccessToken()
        {
            IApplication app = null;
            string access_token = null;

            #region code/csharp/authentication/create_account_linkedin_providerdata_access_token.cs
            var request = client.Providers()
                .LinkedIn()
                .Account()
                .SetAccessToken(access_token)
                .Build();
            var result = await app.GetAccountAsync(request);
            #endregion
        }

        public async Task GetSamlPolicy()
        {
            IApplication app = null;

            #region code/csharp/authentication/saml_policy_example.cs
            var samlPolicy = await app.GetSamlPolicyAsync();

            var samlServiceProvider = await samlPolicy.GetSamlServiceProviderAsync();
            // samlServiceProvider.Href is the SAML Service Provider URL

            var ssoInitiationEndpoint = await samlServiceProvider.GetSsoInitiationEndpointAsync();
            // ssoInitiationEndpoint.Href is the SSO Initiation Endpoint
            #endregion
        }

        public async Task CreateApiKeyForAccount()
        {
            IAccount account = null;

            #region code/csharp/authentication/create_apikey_req.cs
            var newApiKey = await account.CreateApiKeyAsync();
            #endregion
        }

        public async Task DeleteApiKey()
        {
            IApplication app = null;
            string api_key_id = null;

            #region code/csharp/authentication/delete_apikey.cs
            var apiKey = await app.GetApiKeyAsync(api_key_id);
            await apiKey.DeleteAsync();
            #endregion
        }

        public async Task DisableApiKey()
        {
            IApplication app = null;
            string api_key_id = null;

            #region code/csharp/authentication/disable_apikey.cs
            var apiKey = await app.GetApiKeyAsync(api_key_id);
            apiKey.SetStatus(ApiKeyStatus.Disabled);
            await apiKey.SaveAsync();
            #endregion
        }

        public async Task AuthenticateBasicApiRequest()
        {
            IApplication app = null;
            string api_key_id = null;
            string api_key_secret = null;

            #region code/csharp/authentication/authenticate_basic_req.cs
            var apiKeyAuthRequest = new ApiKeyRequestBuilder()
                .SetId(api_key_id)
                .SetSecret(api_key_secret)
                .Build();

            var result = await app.AuthenticateAccountAsync(apiKeyAuthRequest);
            var account = await result.GetAccountAsync();
            #endregion
        }
    }
}
